dart/lib/src/protocol/contexts.dart

```diff
@@ -11,107 +13,170 @@ import 'runtime.dart';
-      json['gpu'] = gpuMap;
-    }
-
-    if (runtimes != null) {
-      if (runtimes.length == 1) {
-        final runtime = runtimes[0];
-        if (runtime != null) {
-          final key = runtime.key ?? 'runtime';
-          json[key] = runtime.toJson();
-        }
-      } else if (runtimes.length > 1) {
-        for (final runtime in runtimes) {
-          if (runtime != null) {
-            var key = runtime.key ?? runtime.name.toLowerCase();
-
-            if (json.containsKey(key)) {
-              var k = 0;
-              while (json.containsKey(key)) {
-                key = '$key$k';
-                k++;
+    forEach((key, value) {
+      if (value == null) return;
+      switch (key) {
+        case Device.type:
+          Map<String, dynamic> deviceMap;
+          if ((deviceMap = device.toJson()).isNotEmpty) {
+            json['device'] = deviceMap;
+          }
+          break;
+        case OperatingSystem.type:
+          Map<String, dynamic> osMap;
+          if ((osMap = operatingSystem.toJson()).isNotEmpty) {
+            json[OperatingSystem.type] = osMap;
+          }
+          break;
+
+        case App.type:
+          Map<String, dynamic> appMap;
+          if ((appMap = app.toJson()).isNotEmpty) {
+            json[App.type] = appMap;
+          }
+          break;
+
+        case 'browser':
+          Map<String, dynamic> browserMap;
+          if ((browserMap = browser.toJson()).isNotEmpty) {
+            json['browser'] = browserMap;
+          }
+          break;
+
+        case 'gpu':
+          Map<String, dynamic> gpuMap;
+          if ((gpuMap = gpu.toJson()).isNotEmpty) {
+            json['gpu'] = gpuMap;
+          }
+          break;
+
+        case 'runtimes':
+          if (runtimes != null) {
+            if (runtimes.length == 1) {
+              final runtime = runtimes[0];
+              if (runtime != null) {
+                final key = runtime.key ?? 'runtime';
+                json[key] = runtime.toJson();
+              }
+            } else if (runtimes.length > 1) {
+              for (final runtime in runtimes) {
+                if (runtime != null) {
+                  var key = runtime.key ?? runtime.name.toLowerCase();
+
+                  if (json.containsKey(key)) {
+                    var k = 0;
+                    while (json.containsKey(key)) {
+                      key = '$key$k';
+                      k++;
+                    }
+                  }
+
+                  json[key] = runtime.toJson()
+                    ..addAll(<String, String>{'type': 'runtime'});
+                }
               }
             }
-
-            json[key] = runtime.toJson()
-              ..addAll(<String, String>{'type': 'runtime'});
           }
-        }
+
+          break;
+
+        default:
+          json[key] = value;
       }
-    }
+    });
 
     return json;
   }
+
+  Contexts clone() {
+    // TODO ? make each class cloneable ?
```

Comment:
is this related to this? https://github.com/getsentry/sentry-dart/projects/1#card-47865684
if so, we can remove the comment, we already have a card for it

Reply:
✅